Fix C balance and FPE errors

peterdschwartz opened the pull request:
Carbon balance was never checked correctly and let through negative values. 
Initialization of Ecosystem variables needed to be re-adjusted, and balance checking for FATES needed to be changed, which causes round-off differences in `CMASS_BALANCE_ERROR` for `fate_cold_allvars` test.

Also included divide-by-zero check in PhenologyMod.F90 to fix fpe in debug mode.

[non-BFB] for one FATES test.
Fixes #6120 
Fixes #6203 
Fixes #6177

peterdschwartz commented:
@glemieux Here is a PR to fix the balance checking issues.  I did get DIFF for `fates_cold_allvars` but only for `CMASS_BALANCE_CHECK` but the values were still `~1E-16`.  

Let me know if how I implemented the new balance checking makes sense to you.

glemieux reviewed (approved):
The implementation looks good to me.  Thanks!

peterdschwartz commented:
After doing a test merge I obtained FILLDIFF for the `fates_cold_allvars`. That I need to look into today

rljacob commented:
What is the diff with?  A baseline from the previous code or is it a diff within a test like an ERS test?

peterdschwartz commented:
Not baseline diff (though that is also there but makes sense to me).  It fails the COMP_base_rest portion of ERP test (same thing with the ERS version I made for debugging).

rljacob commented:
Try a REP test.

peterdschwartz commented:
REP passes just fine. I'm going to double check that each variable for carbon balance is read/written to the restart file, and maybe a longer simulation will reveal diff's accruing in other fields.

bishtgautam commented on components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90:
```diff
@@ -624,7 +621,7 @@ subroutine ColPBalanceCheck(bounds, &
       dt = dtime_mod
       kyr = year_curr; kmo = mon_curr; kda = day_curr; mcsec = secs_curr;
-
+      !nstep = nstep_mod 
```
Please delete this line.

bishtgautam commented on components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90:
```diff
@@ -901,7 +898,14 @@ subroutine GridCBalanceCheck(bounds, col_cs, col_cf, grc_cs, grc_cf)
          grc_som_c_yield           => grc_cf%somc_yield            , & ! Output: [real(r8) (:) ] (gC/m^2/s)total SOM C loss by erosion
          grc_cinputs               => grc_cf%cinputs               , & ! Output: [real(r8) (:) ] (gC/m2/s) column-level C inputs
-         grc_coutputs              => grc_cf%coutputs                & ! Output: [real(r8) (:) ] (gC/m2/s) column-level C outputs
+         grc_coutputs              => grc_cf%coutputs              , & ! Output: [real(r8) (:) ] (gC/m2/s) column-level C outputs
+         beg_totpftc               =>  grc_cs%beg_totpftc          , & ! Input: [real(r8) (:)] (gC/m2) patch-level carbon aggregated to column level, incl veg and cpool
+         beg_cwdc                  =>  grc_cs%beg_cwdc             , & ! Input: [real(r8) (:)] (gC/m2) total column coarse woody debris carbon
+         beg_totsomc               =>  grc_cs%beg_totsomc          , & ! Input: [real(r8) (:)] (gC/m2) total column soil organic matter carbon
+         beg_totlitc               =>  grc_cs%beg_totlitc          , & ! Input: [real(r8) (:)] (gC/m2) total column litter carbon
+         beg_totprodc              =>  grc_cs%beg_totprodc         , & ! Input: [real(r8) (:)] (gC/m2) total column wood product carbon
+         beg_ctrunc                =>  grc_cs%beg_ctrunc           , & ! Input: [real(r8) (:)] (gC/m2) total column truncation carbon sink
+         beg_cropseedc_deficit     =>  grc_cs%beg_cropseedc_deficit  & ! Input: [real(r8) (:)] (gC/m2) column carbon pool for seeding new growth
```
Please remove an extra whitespace after `=>  grc_cs%` to be `=> grc_cs%`

bishtgautam commented on components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90:
```diff
@@ -940,13 +944,21 @@ subroutine GridCBalanceCheck(bounds, col_cs, col_cf, grc_cs, grc_cf)
-               c2l_scale_type = 'unity', l2g_scale_type = 'unity')
+               c2l_scale_type = 'unity', l2g_scale_type = 'unity')  
 
+      if(use_fates) then 
```
please add a whitespace `if (use_fates)`

bishtgautam commented on components/elm/src/biogeochem/EcosystemBalanceCheckMod.F90:
```diff
@@ -940,13 +944,21 @@ subroutine GridCBalanceCheck(bounds, col_cs, col_cf, grc_cs, grc_cf)
       do g = bounds%begg, bounds%endg
-         grc_cinputs(g) = grc_gpp(g) + grc_dwt_seedc_to_leaf(g) + grc_dwt_seedc_to_deadstem(g)
+
+         if(.not. use_fates) then 
```
similarly add a whitespace.

bishtgautam commented on components/elm/src/biogeochem/PhenologyMod.F90:
```diff
@@ -2559,8 +2559,11 @@ subroutine CropPlantDate (num_soilp, filter_soilp, num_pcropp, filter_pcropp, &
+         
          ! calculate the P:PET for each month
-         p2ETo(p,kmo) = xp(p,kmo)/ETo(p,kmo)
+         if( abs(ETo(p,kmo)) > 0._r8) then 
```
- please add whitespace
- Shouldn't there also be code corresponding to an `else` condition?

bishtgautam commented on components/elm/src/data_types/ColumnDataType.F90:
```diff
@@ -7764,6 +7766,8 @@ subroutine col_cf_setvalues ( this, num_column, filter_column, value_column)
        this%litterc_loss(i)              = value_column
        
        this%nee(i)                       = value_column
+       this%er(i)                        = value_column !! Grid Balance Error 
+       this%som_c_leached(i)             = value_column !! Grid Balance Error 
```
Can you please remove the `!! Grid Balance Error`

bishtgautam reviewed (changes requested):
I have minor stylistic code change requests for this PR.

rljacob commented:
@peterdschwartz what is the status of this PR?

peterdschwartz commented:
@rljacob I met with @glemieux  a couple of weeks ago. He mentioned a possible time-step dependent part of FATES that could be the cause, but I think he's been busy with the FATES api update and our follow up meeting with @thorntonpe was cancelled due to lab holiday.  

Today, I'll try increasing the history output to be every timestep and see if that provides any useful information.

peterdschwartz commented:
Update:

Increasing output freq finally showed a new diff in a field.  Looking into it further, it appears that the tendency term for carbon are not added to the leaching term due to some issue 5 years ago (according to git blame atleast), but it is for nitrogen and phosphorus. 

Hopefully the straightforward solution works here.
```
 LITR2C_TNDNCY_VERT_TRANS   (lndgrid,levdcmp,time)  t_index =    194   194
          3     5760  (   153,     2,     1) (   153,     1,     1) (   237,     3,     1) (   247,    10,     1)
                3165   1.043116371057806E-09  -1.643725044964128E-09 2.2E-19  2.944499461962624E-12 8.1E-11  1.293930177203629E-16
                3165   1.043116371057806E-09  -1.643725044964128E-09          2.944499678803059E-12          1.293930309552527E-16
                5760  (   153,     2,     1) (   153,     1,     1)
          avg abs field values:    4.034842251976478E-11    rms diff: 0.0E+00   avg rel diff(npos):  8.1E-11
                                   4.034842251976478E-11                        avg decimal digits(ndif):  7.1 worst:  7.0
 RMS LITR2C_TNDNCY_VERT_TRANS         0.0000E+00            NORMALIZED  0.0000E+00
```

glemieux commented:
@peterdschwartz can you point me to the line of code with the git blame that you noted earlier, please?

peterdschwartz commented:
@glemieux This [Code block](https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/data_types/ColumnDataType.F90#L7531) is what I was referring to. The transport tendency is not added to carbon leached pool even though the corresponding terms are added for Nitrogen and Phosphorus.  

More experimenting showed that bypassing the SoilLittVertTransp calculation, the restart failure goes away. For FATES, the `UpdateLitterFluxes`  is done just prior that adjusts the inputs to the `SoilLittVertTransp` that could be the cause.  [Line](https://github.com/E3SM-Project/E3SM/blob/master/components/elm/src/biogeochem/EcosystemDynMod.F90#L706)

peterdschwartz commented:
Since there's a previous fates related issue linked above with the same errors, I'm going to just clean the branch up for the other issues and follow the review comments. The other issue can be used to the issue.

glemieux commented:
> Since there's a previous fates related issue linked above with the same errors, I'm going to just clean the branch up for the other issues and follow the review comments. The other issue can be used to the issue.

Sorry for the delay in trying to address the fates-side of this.  Just to clarify, you're saying your going to move forwards towards integration @peterdschwartz?  If so, that sounds good to me.

peterdschwartz commented:
> > Since there's a previous fates related issue linked above with the same errors, I'm going to just clean the branch up for the other issues and follow the review comments. The other issue can be used to the issue.
> 
> Sorry for the delay in trying to address the fates-side of this. Just to clarify, you're saying your going to move forwards towards integration @peterdschwartz? If so, that sounds good to me.

@glemieux  Yes, and now I know how to how to reliably reproduce the issue, we can use #6125 to track progress. I will share a branch with the changes to do it.

@bishtgautam I have rebased the branch and addressed your comments.  There's potentially a couple of different ways to handle the `else` condition in the PhenologyMod, so I went with what I believe to be the simplest and added a comment explaining why.  Let me know what you think

e3sm_developer came back [BfB] with the exception of one FATES test that will FAIL restart comparison for the time being.

bishtgautam commented on components/elm/src/biogeochem/PhenologyMod.F90:
```diff
@@ -2559,8 +2559,17 @@ subroutine CropPlantDate (num_soilp, filter_soilp, num_pcropp, filter_pcropp, &
+         
          ! calculate the P:PET for each month
-         p2ETo(p,kmo) = xp(p,kmo)/ETo(p,kmo)
+         if ( abs(ETo(p,kmo)) > 0._r8) then 
+            p2ETo(p,kmo) = xp(p,kmo)/ETo(p,kmo)
+         else ! P:PET is undefined.
+            ! Setting to a fill value ( 'spval' ) would
+            ! require nested if statements due to 
+            ! the weighting of previous years (i.e., p2ETo and prev_p2ETo_bar )
+            ! So, set to zero for simplicity.
+            p2ETo(p,kmo) = 0._r8
+         end if 
```
@peterdschwartz This looks good to me. Thanks for the comment explaining the logic.

rljacob commented:
This is ready to merge.
